tvm/src/codegen/llvm/codegen_llvm.cc

```diff
@@ -1588,6 +1610,92 @@ void CodeGenLLVM::VisitStmt_(const MultiBlock* op) {
   }
 }
 
+void CodeGenLLVM::VisitStmt_(const Assert* op) {
+  std::vector<llvm::Value*> values;
+  std::vector<Type> types;
+  std::vector<llvm::Type*> llvm_types;
+  for (size_t i = 0; i < op->values.size(); i++) {
+    Expr v = op->values[i];
+    values.push_back(MakeValue(v));
+    types.push_back(v.type());
+    if (v.type().is_int() || v.type().is_uint()) {
+      llvm_types.push_back(t_int64_);
+    } else {
+      llvm_types.push_back(llvm::Type::getDoubleTy(*ctx_));
+    }
+  }
+  llvm::FunctionType* call_ftype = llvm::FunctionType::get(t_int_, true);
+#if TVM_LLVM_VERSION <= 60
+  llvm::Function* printf_call = llvm::cast<llvm::Function>(
+      module_->getOrInsertFunction("printf", call_ftype));
+#else
+  llvm::Function* printf_call = llvm::cast<llvm::Function>(
+      module_->getOrInsertFunction("printf", call_ftype).getCallee());
+#endif
+
+  std::vector<llvm::Value*> printf_args;
+  std::string message = op->message;
+  printf_args.push_back(builder_->CreateGlobalStringPtr(message));
+
+  for (size_t i = 0; i < op->values.size(); i++) {
+    if (types[i].is_int() || types[i].is_uint()) {
+      llvm::Value* ivalue = CreateCast(types[i], Int(64), values[i]);
+      printf_args.push_back(ivalue);
+    } else {  // fixed or float
+      llvm::Value* fvalue = CreateCast(types[i], Float(64), values[i]);
+      printf_args.push_back(fvalue);
+    }
+  }
+  using llvm::BasicBlock;
+
+  size_t num_free = assert_alloc_mem.size();
+  std::string if_then_name = "if_then_";
+  std::string if_end_name = "if_end_";
+
+  llvm::Value* cond = MakeValue(op->condition);
+
+  BasicBlock* assertstmt_true =
+      BasicBlock::Create(*ctx_, "assertstmt_true", function_);
+  BasicBlock* assertstmt_false =
+      BasicBlock::Create(*ctx_, "assertstmt_false", function_);
+  BasicBlock* end_assert_here =
+      BasicBlock::Create(*ctx_, "end_assert_here", function_);
```

Comment:
We don't really need `end_assert_here`, right? We just need to set insert point to `assertstmt_true` at the end.